pkg/yang/types_builtin.go

```diff
@@ -588,17 +696,36 @@ func (n Number) Less(m Number) bool {
 		return true
 	case n.Kind != Negative && m.Kind == Negative:
 		return false
-	case n.Kind == Negative:
-		return n.Value > m.Value
+	}
+
+	// Normalize into a common representation dictated by the largest number of
+	// fraction digits in either number. Values can be compared directly in this
+	// representation.
+	maxfd := max(n.FractionDigits, m.FractionDigits)
+	nd, md := n.toDecimal(maxfd), m.toDecimal(maxfd)
+
+	switch n.Kind {
+	case Negative:
+		return nd.Value > md.Value
+	case Positive:
+		return nd.Value < md.Value
 	default:
-		return n.Value < m.Value
 	}
+	panic("Less with unknown number type")
 }
 
-// Equal returns true if m equals n.  It provides symmetry with the Less
-// method.
+// Equal returns true if n is equal to m.
 func (n Number) Equal(m Number) bool {
-	return n == m
+	return !n.Less(m) && !m.Less(n)
+}
+
+// toDecimal returns n represented internally with fracDigits fractional digits.
+func (n Number) toDecimal(fracDigits uint8) Number {
+	if fracDigits > MaxFractionDigits {
+		return Number{Kind: n.Kind, Value: 0, FractionDigits: MaxFractionDigits}
+	}
+
+	return Number{Kind: n.Kind, Value: n.Value * uint64(math.Pow10(int(fracDigits-n.FractionDigits))), FractionDigits: fracDigits}
```

Comment:
This will always set the value to 0 if n.FractionDigits is greater than fracDigits.  You need to use the absolute value and then divide in that case.  See https://play.golang.org/p/Bz0jsgj0ae

Please first add tests that fail with the existing code and then fix the code so the tests pass.